packages/core/contracts/financial-templates/common/financial-product-libraries/KpiOptionsFinancialProductLibrary.sol

```diff
@@ -11,6 +11,42 @@ import "../../../common/implementation/Lockable.sol";
  * Post-expiry, the collateral requirement is left as 1 and the price is left unchanged.
  */
 contract KpiOptionsFinancialProductLibrary is FinancialProductLibrary, Lockable {
+    using FixedPoint for FixedPoint.Unsigned;
+
+    mapping(address => FixedPoint.Unsigned) financialProductTransformedPrices;
+
+    /**
+     * @notice Enables any address to set the transformed price for an associated financial product.
+     * @param financialProduct address of the financial product.
+     * @param transformedPrice the price for the financial product to be used if the contract is pre-expiration.
+     * @dev Note: a) Any address can set price transformations b) The price can't be set to blank.
+     * c) A transformed price can only be set once to prevent the deployer from changing it after the fact.
+     * d)  financialProduct must expose an expirationTimestamp method.
+     */
+    function setFinancialProductTransformedPrice(address financialProduct, FixedPoint.Unsigned memory transformedPrice)
+        public
+        nonReentrant()
+    {
+        require(transformedPrice.isGreaterThan(0), "Cant set price of 0");
+        require(financialProductTransformedPrices[financialProduct].isEqual(0), "Price already set");
+        require(ExpiringContractInterface(financialProduct).expirationTimestamp() != 0, "Invalid EMP contract");
```

Comment:
Think you should move this external function call to the FinancialContract after the state change to stick to the [Check-Effects-Interaction pattern](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html). But curious what @chrismaree @mrice32 think too

Reply:
interesting. I see this more as a _check_ part of the pattern as this is not actually an _interaction_. it's validating that the `financialProduct` is an EMP indirectly by ensuring it has this `expirationTimestamp` method` and is not actually interacting with the external contract.

More over, I personally prefer having all the requires at the top of the method. Lastly, this method has `nonReentrant` modifier so nothing negative can come from having it above the state modification.

Reply:
But couldn't you theoretically set a malicious `expirationTimestampI()` implementation that tries to do something else in addition to returning the expiration time?

Reply:
If this library had any token balance in it or something like that then I guess you are right from that perspective. I think that having the modifier is sufficient though, don't you?

Reply:
yeah good call, forgot about `nonReentrant`